src/main/java/org/jabref/logic/citationstyle/CitationStyleCache.java

```diff
@@ -65,8 +65,8 @@ public void listen(EntryChangedEvent entryChangedEvent) {
          * removes the citation of the removed entry as it's not needed anymore
          */
         @Subscribe
-        public void listen(EntryRemovedEvent entryRemovedEvent) {
-            citationStyleCache.invalidate(entryRemovedEvent.getBibEntry());
+        public void listen(EntriesRemovedEvent entriesRemovedEvent) {
+            citationStyleCache.invalidateAll(entriesRemovedEvent.getBibEntries());
```

Comment:
I wouldn't change to many functions to accept multiple elements and only if there are (performance) advantages. Here for example a simple for each loop should have been sufficient.

Reply:
Okay. In my changes, I replaced `EntryRemovedEvent` with `EntriesRemovedEvent` everywhere. Should I put it back? The reason I changed the method was in order to change the following method to only send one event.
https://github.com/JabRef/jabref/blob/b08ae88d6f80585dcb34b450217699b6b6104da4/src/main/java/org/jabref/model/database/BibDatabase.java#L256-L269

Reply:
The changed events are fine and a good way of doing things. What I meant is that this change then does not need to propagate further and you can use for-each-loops in the event handler. For example, removing many entries from the main table should be done in one pass (so that the ui is only redrawn once) but invalidating the citation style cache can be done in a foreach loop since this does not (really) profit from the batch operation. Probably needs to be decided case-by-case.

Reply:
Okay. I'm not familiar with a lot of the code where the events propagate. I could try and run the GUI and try to reach all the different cases where events are handled to figure out performance bottlenecks, but that would take a long time because there are so many different listeners. Alternatively, I could just do for loops for most or all of them except for the ones in `DBMSSychronizer`, where the events clearly need to correspond to batch SQL queries. Do you have any suggestions on this?

Another issue that is coming up is that I am not an expert on SQL, especially not MySQL and Oracle. Is this the kind of thing I can learn from documentation or will I need help?

Reply:
@NorwayMaple  A good way to analyze performance bottlenecks is using a profiler, for example we use the Yourkit JavaProfiler: https://www.yourkit.com/features/ (you can get a 15 days trial key)

Regarding the Oracle/MySQL stuff: I think nobody of us has really deep insight. I would stick with the official documentation. Most should be related to syntax

Reply:
I agree, the database improvement should be the highest priority. However, since I have gone ahead and changed so many of the classes that involve `EntryEvent`s, it won't compile until I finish at least some work there. From my understanding, I also would have to make sure that everywhere that called `BibDatabase.removeEntry` iteratively (within a for loop) changed to calling `BibDatabase.removeEntries` once. Also, this would mean taking everywhere that listened for an `EntryRemovedEvent` and replacing it with an `EntriesRemovedEvent`. Otherwise, this defeats the purpose of `removeEntries` and it will be impossible to test the database improvements. Is there any way to make this easier?

Reply:
No, I'm afraid there is no easier way than what you describe...